### DISCUSSION TITLE
k3s server create 100% CPU load

### DISCUSSION THREAD

**Discussion by allnightlong** [k3s-io/k3s#11251] [Q&A] [answered]
**Environmental Info:**
K3s Version: 
k3s -v
k3s version v1.30.6+k3s1 (1829eaae)
go version go1.22.8

Node(s) CPU architecture, OS, and Version: 
Linux ***-master 6.8.0-48-generic #48-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 27 14:04:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
1 server 
6 agents

**Describe the bug:**
100% CPU usage of k3s processes on master node. Can't diagnose and find out root of the problem.
HTOP screenshot
<img width="2017" alt="Screenshot 2024-11-06 at 18 19 12" src="https://github.com/user-attachments/assets/f7174d28-76a7-4524-b7f2-ae78d23108d8">

NewRelic CPU usage for 3 days screenshot
<img width="1855" alt="Screenshot 2024-11-06 at 18 20 41" src="https://github.com/user-attachments/assets/2158cfb4-c1ab-44fe-b415-fc960f55043d">

**Steps To Reproduce:**
I'm not sure exactly, how to reproduce the problem. My application creates a lot of short-living kube jobs. 
Could it be the cleanup problem?

As 3 days graph suggests, the CPU load increases day-by-day (while the application is having the same load), so `k3s` finally arrives at the point, where CPU load is unbearable. 

I've tried to run:
```
k3s-killall.sh
systemctl start k3s
```
from this topic https://github.com/k3s-io/k3s/issues/6095#issuecomment-1305282469, but it never helped.

So I've ended up reinstalling the whole cluster on my new machine, just to find myself in the same situation 5 days after.

**Expected behavior:**
CPU load shouldn't reach over 100%

**Actual behavior:**
CPU load reaches 100%, and as a result it takes too long for kubernetes to launch pods, so the jobs falls out of the deadline.

**Additional context / logs:**
I can provide any necessary logs on the request

**Comment by brandond** [comment]
I'd probably look at what the cluster is doing? Check the logs for errors, do `kubectl events -A -w` to see if there is excessive thrashing from any of the controllers, and so on? Without knowing what's causing the load it's hard to suggest any specific actions.

The fact that CPU utilization rises and falls over time suggests that the load is related to external factors, and is not just an issue with K3s itself.

**Reply by allnightlong** [reply]
Agreed. 

It's also very difficult for me to understand the root of the problem. The cpu load indeed correlate to the application load, but it raises over the time non-proportionally (avg cpu usage is 10% on night 1, 20% - night 2, 50% - night 3).

For me, it's an indicator of spending CPU to some kind of garbage collection problem.
 ```
kubectl events -A -w
```
doesn't show anything criminal. It's a lot of `job execution` logs:
```
8h                      Normal    Scheduled                        Pod/media-yt-jfpqv-twoug-399-140-54ndj                        Successfully assigned ***/media-yt-jfpqv-twoug-399-140-54ndj to ***-node-nu
8h                      Normal    Pulled                           Pod/media-yt-jfpqv-twoug-399-140-54ndj                        Container image "jauderho/yt-dlp:2024.10.22" already present on machine
8h                      Normal    Created                          Pod/media-yt-jfpqv-twoug-399-140-54ndj                        Created container media-yt-jfpqv-twoug-399-140
8h                      Normal    Started                          Pod/media-yt-jfpqv-twoug-399-140-54ndj                        Started container media-yt-jfpqv-twoug-399-140
8h                      Normal    Completed                        Job/media-yt-jfpqv-twoug-399-140                              Job completed
```


I've also check journal, and didn't find anything criminal.
Could you suggest anything else?

**Reply by brandond** [reply]
And there's nothing interesting in the logs?

You could try enable profiling on the server node and then grab a cpu profile, when it's at a period of high resource usage.

**Comment by allnightlong** [comment]
Sorry, I'm not that familiar with cpu profiling. Could you suggest any tool to do that?

**Reply by brandond** [reply]
1. start k3s server with `--enable-pprof`
2. save a profile: `kubectl get --raw /debug/pprof/profile?seconds=60 > cpu.profile.gz`
3. Upload the file here

**Reply by allnightlong** [reply]
I've updated file `/etc/systemd/system/k3s.service` with `--enable-pprof`:
```
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server \
        '--tls-san' \
        'master.xxx.cloud' \
        '--node-external-ip=x.x.x.x' \
        '--flannel-backend=wireguard-native' \
        '--flannel-external-ip' \
        '--bind-address=0.0.0.0' \
        '--kubelet-arg=allowed-unsafe-sysctls=net.ipv6.*' \
        '--kubelet-arg=allowed-unsafe-sysctls=net.ipv4.*' \
        '--enable-pprof' \
```
then restarted `k3s`:
```
$ sudo service k3s restart
```
But when I try to get data, I've got error
```
$ kubectl get --raw /debug/pprof/profile?seconds=60                                                                   
zsh: no matches found: /debug/pprof/profile?seconds=60
```

**Reply by brandond** [reply]
you got an error from zsh not kubectl so I don't know what's going on there. Did you need to quote the url or something? I don't personally use zsh.

Also note that you will want to redirect that output into a .gz file as it is a bunch of binary that you do NOT want dumped to your shell.

**Reply by allnightlong** [reply]
I've managed to resolve `zsh` issues, but I still got an error:
```
$ kubectl get --raw /debug/pprof/profile?seconds=60
Error from server (NotFound): the server could not find the requested resource
```

Should I add pod or service or smth else here?

**Reply by brandond** [reply]
Sounds like the flag didn't get added. Did you do a `systemctl daemon-reload` after modifying the systemd unit, before restarting the service? You can also check the `k3s` process in `ps` and confirm that the flag is there.

**Reply by allnightlong** [reply]
Thank you for your patience, @brandond - I've finally managed to profile the system.
For possible future readers:
Just modifying the `/etc/systemd/system/k3s.service` and then restarting the service `sudo service k3s restart` is not enough.
I had to reconfigure `k3s` from scratch, adding `--enable-pprof` in the end of `INSTALL_K3S_EXEC`. Example: 
```
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC='server --tls-san core.xxx.cloud --node-external-ip=a.b.c.d --flannel-backend=wireguard-native --flannel-external-ip --bind-address=0.0.0.0  --kubelet-arg=allowed-unsafe-sysctls=net.ipv6.* --kubelet-arg=allowed-unsafe-sysctls=net.ipv4.* --enable-pprof' INSTALL_K3S_CHANNEL='stable' sh -
```

Here is my profiling file:
[cpu.profile.gz](https://github.com/user-attachments/files/17669309/cpu.profile.gz)

**Reply by brandond** [reply]
That's not necessary, but it is another way to achieve the same result. A common mistake when modifying the systemd unit, other than not reloading before restarting, is leaving a trailing `\` at the end of the last arg, but not following it with a blank line.

**Comment by allnightlong** [comment]
Btw, I've removed all the agent and load from the server, but it still consumes about 50% of CPU
![2024-11-07_02-50-45](https://github.com/user-attachments/assets/ce96080b-cb5f-4f6d-a38f-cec2b037fefa)

**Comment by brandond** [comment]
Pretty much the ONLY thing in your trace is sqlite:
![image](https://github.com/user-attachments/assets/2681ff65-8d75-4ae9-b34c-0a09511d529b)

I can't say I've ever seen this before, usually sqlite is very fast. What sort of storage are you running this on? Check journald for messages in your k3s service logs about compact - is it failing to compact the datastore for some reason?

**Reply by allnightlong** [reply]
Here is my VM config. I use High-IOPS sdd, so I don't expect the problem from this point of view.
![2024-11-08_01-39-05](https://github.com/user-attachments/assets/3891e962-5673-46b8-8644-34fe326f02f1)
![2024-11-08_01-38-01](https://github.com/user-attachments/assets/c1eb264b-530b-4cfd-9f17-b63830e60079)

As for journal, I'm using `journalctl -xeu k3s.service` to check, and it shows only last 1000 entries. They don't contain anything related to `compact`.

My journal looks like this:
![2024-11-08_01-42-53](https://github.com/user-attachments/assets/212a6831-eb93-4198-a6b8-3e7e16c62821)
Is there any other way to check logs?

**Reply by brandond** [reply]
You're enqueueing 10s of jobs per second? That seems... excessive. No wonder you're putting so much pressure on the datastore. Is there a problem here, or is this how it normally looks?

Can you dump the logs with `journalctl -u k3s | gzip -c > k3s.log.gz` and attach the logs here?

**Reply by allnightlong** [reply]
I've replaced my real cluster name with `k3s-cluster`, everything else is intact.
[k3s.log.gz](https://github.com/user-attachments/files/17669931/k3s.log.gz)

**Reply by allnightlong** [reply]
And yes, as I've said in the begging of the thread: `My application creates a lot of short-living kube jobs.`, so it's a normal working process for my cluster.

**Comment by brandond** [comment, marked as answer]
OK yeah so your compact is regularly failing with `database is locked` errors:
```
Nov 07 21:34:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:49Z" level=info msg="COMPACT compactRev=1170296 targetCompactRev=1171296 currentRev=2105439"
Nov 07 21:34:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:49Z" level=info msg="COMPACT deleted 846 rows from 1000 revisions in 337.232541ms - compacted to 1171296/2105439"
Nov 07 21:34:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:49Z" level=info msg="COMPACT compactRev=1171296 targetCompactRev=1172296 currentRev=2105442"
Nov 07 21:34:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:49Z" level=info msg="COMPACT deleted 1050 rows from 1000 revisions in 248.221705ms - compacted to 1172296/2105442"
Nov 07 21:34:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:49Z" level=info msg="COMPACT compactRev=1172296 targetCompactRev=1173296 currentRev=2105448"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT deleted 915 rows from 1000 revisions in 236.738881ms - compacted to 1173296/2105448"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT compactRev=1173296 targetCompactRev=1174296 currentRev=2105452"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT deleted 948 rows from 1000 revisions in 236.723365ms - compacted to 1174296/2105452"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT compactRev=1174296 targetCompactRev=1175296 currentRev=2105457"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT deleted 1059 rows from 1000 revisions in 235.299717ms - compacted to 1175296/2105457"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT compactRev=1175296 targetCompactRev=1176296 currentRev=2105464"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT deleted 989 rows from 1000 revisions in 259.278141ms - compacted to 1176296/2105464"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=info msg="COMPACT compactRev=1176296 targetCompactRev=1177296 currentRev=2105470"
Nov 07 21:34:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:34:50Z" level=error msg="Compact failed: failed to compact to revision 1177296: database is locked"

Nov 07 21:39:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:39:49Z" level=info msg="COMPACT compact revision changed since last iteration: 1170296 => 1176296"

Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT compactRev=1176296 targetCompactRev=1177296 currentRev=2115228"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT deleted 1032 rows from 1000 revisions in 232.766519ms - compacted to 1177296/2115228"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT compactRev=1177296 targetCompactRev=1178296 currentRev=2115229"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT deleted 961 rows from 1000 revisions in 237.813295ms - compacted to 1178296/2115229"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT compactRev=1178296 targetCompactRev=1179296 currentRev=2115229"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT deleted 932 rows from 1000 revisions in 169.745804ms - compacted to 1179296/2115229"
Nov 07 21:44:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:49Z" level=info msg="COMPACT compactRev=1179296 targetCompactRev=1180296 currentRev=2115229"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT deleted 958 rows from 1000 revisions in 207.300242ms - compacted to 1180296/2115229"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT compactRev=1180296 targetCompactRev=1181296 currentRev=2115235"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT deleted 1050 rows from 1000 revisions in 212.96679ms - compacted to 1181296/2115235"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT compactRev=1181296 targetCompactRev=1182296 currentRev=2115236"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT deleted 987 rows from 1000 revisions in 119.917869ms - compacted to 1182296/2115236"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT compactRev=1182296 targetCompactRev=1183296 currentRev=2115236"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT deleted 1032 rows from 1000 revisions in 91.848726ms - compacted to 1183296/2115236"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=info msg="COMPACT compactRev=1183296 targetCompactRev=1184296 currentRev=2115240"
Nov 07 21:44:50 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:44:50Z" level=error msg="Compact failed: failed to compact to revision 1184296: database is locked"

Nov 07 21:49:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:49:49Z" level=info msg="COMPACT compact revision changed since last iteration: 1176296 => 1183296"

Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT compactRev=1183296 targetCompactRev=1184296 currentRev=2125744"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT deleted 1048 rows from 1000 revisions in 153.422786ms - compacted to 1184296/2125744"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT compactRev=1184296 targetCompactRev=1185296 currentRev=2125747"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT deleted 906 rows from 1000 revisions in 216.277176ms - compacted to 1185296/2125747"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT compactRev=1185296 targetCompactRev=1186296 currentRev=2125753"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT deleted 950 rows from 1000 revisions in 191.92297ms - compacted to 1186296/2125753"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=info msg="COMPACT compactRev=1186296 targetCompactRev=1187296 currentRev=2125754"
Nov 07 21:54:49 k3s-cluster-core k3s[2668037]: time="2024-11-07T21:54:49Z" level=error msg="Compact failed: failed to compact to revision 1187296: database is locked"
```

When the compact only partially completes, it skips the next cycle and then there is even more work to do the next time, so it is less likely to catch up. This also leaves more data in the database for it to sort through, which is probably the cause of your excessive load.

You might consider switching to etcd? If you have a single server, you can just convert in-place by adding `--cluster-init` to your args.

While we can try to handle the "database is locked" issue better, sqlite is really tuned for smaller use cases where low overhead and low disk IO is more important than raw throughput. Etcd will probably suit your use case better.

**Reply by allnightlong** [reply]
Thank you for your answer, @brandond , I'm really appreciating it.

I've switched my cluster to `etcd` and CPU load immediately dropped.
I'll continue to monitor for the issue, and come back to this thread with the results.

**Reply by allnightlong** [reply]
![2024-11-08_03-20-33](https://github.com/user-attachments/assets/d848dfe8-91bd-4577-bf14-8d2bff84cc4b)
Here is newrelic stats for etcd update

**Comment by brandond** [comment]
@allnightlong what filesystem and disk type are you running k3s on? I am trying to reproduce the `database is locked` errors but I have been unable to do so even under significant load.

**Reply by allnightlong** [reply]
@brandond 

I've already provided you the [information](https://github.com/k3s-io/k3s/discussions/11251#discussioncomment-11182646) from my hosting company. It's positioned as High-IOPS.

I also can run `lsblk` on the machine.
```
lsblk -o PATH,SIZE,FSTYPE,MOUNTPOINT
PATH        SIZE FSTYPE MOUNTPOIN
/dev/vda    200G
/dev/vda1   199G ext4   /
/dev/vda14    4M
/dev/vda15  106M vfat   /boot/efi
/dev/vda16  913M ext4   /boot
```

If you need any extra information, feel free to ask - I'll run the command required.

**Comment by serhiynovos** [comment]
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/a573eea4-09c5-4b0e-b485-9a3913ed54cc">

I noticed something similar on two of my k3s clusters. One is single node only with rancher installed. Second one (what you see on the screenshot) CPU started increasing after I installed longhorn on it and installed my helm chart with about 18 pods.  So my clusters are pretty small and nothing is enqueued constantly just some peak load when something was installed. When pods stopped scheduling and all got stalled shouldn't CPU usage be normalized back ?

Just restarting k3s service helped to reduce CPU usage

**Reply by serhiynovos** [reply]
These two cluster are running on not very powerful hardware. One with SSD another on HDD raid array

**Comment by serhiynovos** [comment]
@brandond Just noticed I have one more time the same issue with staled single node cluster where I have only rancher running.
<img width="1314" alt="image" src="https://github.com/user-attachments/assets/c6199e44-e002-4f5f-ac88-937b59dba258">

**Reply by serhiynovos** [reply]
Disk IO is the same, about 150kb/s

**Reply by serhiynovos** [reply]
```
Dec 02 18:28:59 rancher k3s[752]: time="2024-12-02T18:28:59Z" level=error msg="error while range on /registry/events/ /registry/events/: rpc error: code = OutOfRange desc = etcdserver: mvcc: required revision has been compacted"
Dec 02 18:29:28 rancher k3s[752]: I1202 18:29:28.276462     752 range_allocator.go:241] "Successfully synced" key="rancher"
Dec 02 18:30:30 rancher k3s[752]: time="2024-12-02T18:30:30Z" level=info msg="COMPACT compactRev=9286906 targetCompactRev=9287537 currentRev=9288537"
Dec 02 18:30:30 rancher k3s[752]: time="2024-12-02T18:30:30Z" level=info msg="COMPACT deleted 630 rows from 631 revisions in 34.757514ms - compacted to 9287537/9288537"
Dec 02 18:30:34 rancher k3s[752]: time="2024-12-02T18:30:34Z" level=info msg="Slow SQL (started: 2024-12-02 18:30:30.9023556 +0000 UTC m=+251714.876574201) (total time: 3.681291281s): PRAGMA wal_checkpoint(FULL)"
Dec 02 18:30:34 rancher k3s[752]: time="2024-12-02T18:30:34Z" level=info msg="Slow SQL (started: 2024-12-02 18:30:32.016580734 +0000 UTC m=+251715.990799321) (total time: 2.638424158s): INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) values(?, ?, ?, ?, ?, ?, ?, ?)"
Dec 02 18:30:34 rancher k3s[752]: time="2024-12-02T18:30:34Z" level=info msg="Slow SQL (started: 2024-12-02 18:30:30.940493419 +0000 UTC m=+251714.914712005) (total time: 3.741514566s): INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) values(?, ?, ?, ?, ?, ?, ?, ?)"
Dec 02 18:30:34 rancher k3s[752]: time="2024-12-02T18:30:34Z" level=info msg="Slow SQL (started: 2024-12-02 18:30:32.565620949 +0000 UTC m=+251716.539839553) (total time: 2.137145095s): INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) values(?, ?, ?, ?, ?, ?, ?, ?)"
Dec 02 18:30:34 rancher k3s[752]: time="2024-12-02T18:30:34Z" level=info msg="Slow SQL (started: 2024-12-02 18:30:32.266937462 +0000 UTC m=+251716.241156066) (total time: 2.44021296s): INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) values(?, ?, ?, ?, ?, ?, ?, ?)"
Dec 02 18:30:44 rancher k3s[752]: time="2024-12-02T18:30:44Z" level=error msg="Failed to list /registry/events/ for revision 9287119: rpc error: code = OutOfRange desc = etcdserver: mvcc: required revision has been compacted"
Dec 02 18:30:44 rancher k3s[752]: W1202 18:30:44.652552     752 watcher.go:338] watch chan error: etcdserver: mvcc: required revision has been compacted
Dec 02 18:30:45 rancher k3s[752]: time="2024-12-02T18:30:45Z" level=error msg="error while range on /registry/events/ /registry/events/: rpc error: code = OutOfRange desc = etcdserver: mvcc: required revision has been compacted"
Dec 02 18:31:08 rancher k3s[752]: time="2024-12-02T18:31:08Z" level=error msg="Failed to list /registry/events/ for revision 9286414: rpc error: code = OutOfRange desc = etcdserver: mvcc: required revision has been compacted"
Dec 02 18:31:08 rancher k3s[752]: W1202 18:31:08.479295     752 watcher.go:338] watch chan error: etcdserver: mvcc: required revision has been compacted
Dec 02 18:31:09 rancher k3s[752]: time="2024-12-02T18:31:09Z" level=error msg="error while range on /registry/events/ /registry/events/: rpc error: code = OutOfRange desc = etcdserver: mvcc: required revision has been compacted"
```

Logs seeing right now

**Reply by serhiynovos** [reply]
One think I noticed that my VM disk was on RAID 10 with HDD and `fio --rw=write --ioengine=sync --fdatasync=1 --directory=test-data --size=100m --bs=2300 --name=mytest` was giving me about 49 IOPS. I moved vm disk to mirrored pool and I see right now about 1050 IOPS and p99 is 1614ms. 


Regarding second cluster: it's just running for test purposes on old mini box with proxmox and cheap SSD. Also I see it was giving small amount of IOPS (about 70). I'm going to move to another newer mini pc where I see 700 IOPS and see how it will go.
